controllers/onload_controller.go

```diff
@@ -603,23 +721,62 @@ func (r *OnloadReconciler) evictOnloadedPods(ctx context.Context, node corev1.No
 	return &ctrl.Result{RequeueAfter: defaultRequeueTime}, nil
 }
 
+type getKernelMapping func(onloadv1alpha1.OnloadKernelMapping) *kmm.KernelMapping
+
+type moduleConfig struct {
+	modprobeArg string
+	moduleName  string
+
+	getKernelMapping getKernelMapping
+}
+
+func getOnloadKernelMapping(spec onloadv1alpha1.OnloadKernelMapping) *kmm.KernelMapping {
+	return &kmm.KernelMapping{
+		Regexp:         spec.Regexp,
+		ContainerImage: spec.KernelModuleImage,
+	}
+}
+
+func getSFCKernelMapping(spec onloadv1alpha1.OnloadKernelMapping) *kmm.KernelMapping {
+	// If the SFC field is not provided, the controller doesn't manage
+	// the SFC kernel module.
+	if spec.SFC == nil {
+		return nil
+	}
+
+	// Otherwise, it reuses the Onload image.
+	return getOnloadKernelMapping(spec)
+}
+
+// Return true if any of the kernel mappings contain a non-nil SFC field.
+func onloadUsesSFC(onload *onloadv1alpha1.Onload) bool {
+	for _, kmapSpec := range onload.Spec.Onload.KernelMappings {
+		kmap := getSFCKernelMapping(kmapSpec)
+		if kmap != nil {
+			return true
+		}
+	}
+
+	return false
+}
+
 func (r *OnloadReconciler) createAndAddModules(
 	ctx context.Context,
 	onload *onloadv1alpha1.Onload,
 ) (*ctrl.Result, error) {
 	log := log.FromContext(ctx)
 
-	moduleNames := []string{
-		onload.Name + onloadModuleNameSuffix,
-		// onload.Name + "-sfc-module",
+	moduleConfigs := []moduleConfig{
+		{"onload", onload.Name + onloadModuleNameSuffix, getOnloadKernelMapping},
+		{"sfc", onload.Name + sfcModuleNameSuffix, getSFCKernelMapping},
```

Comment:
Have you thought about adding a check with `onloadUsesSFC` here to conditionally add the module config to the slice. It could simplify the for loop and `createModule` since we would only enter them when we actually want to do something and so could remove the some of the `== nil` or `len(...) == 0` checks.

Reply:
Yes, it's a good idea. Also, given that it means no changes in `createModule()`, we could remove a new unit test for `== nil` and `len(...) == 0` without sacrificing the test coverage because the optional SFC module kind creation is tested with EnvTests.